Challenge 1: Largest Palindrome
A palindromic number reads the same both ways. For example, 1234321 is a palindrome. The largest palindrome made from the product of two two-digit numbers is 9009 = 91 × 99. Find the largest palindrome made from the product of two three-digit numbers. Afterward, write a brief explanation walking through your code's logic in markdown.


In [163]:
def palindrome():
    list_of_palindrome = []  
    for x in range(100,1000):
        for y in range(100,1000):
            if str(x*y) == str(x*y)[::-1]:
                list_of_palindrome.append(int(x*y))
    print(max(list_of_palindrome)) 

palindrome() 

906609


Find the palindrome numbers and list the largest
Create empty list to add palindrom numbers.
To find the three digit of palindrome, run the loop of all three digit from 100 to 999
When the first for loop runs, x will print number 100 and second loop will run from 100 to 999.
If multiple of two numbers (X * Y) and their reverse numbers [::-1] are equal, numbers aree palindrome. To reverse the number, change the number to string. Add the str(number) to the list and change back to integer.
check the maximum
call a functon
906609 is multipe of 913 *993




Challenge 2: Summation of Primes
The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17. Find the sum of all the primes below 2,000. Afterward, write a brief explanation walking through your code's logic in markdown.

In [164]:
def isPrime (number):
    if number != 1:         
        for f in range(2,number): 
            if number %  f == 0:     
                return False           
    else: 
        return False
    return True
# sum prime numbers of range from 2 below 2000
sum = 0
for f in range(2,2000):
    if isPrime(f) == True:
        sum += f
print (sum)

277050


Defined prime numbers. 
First 'if statement' shows that 1 is not a prime number. Therefor nummbers are 2.3.4.5.6....
Using for loop, search the number divided by a factor 2.3.4.5.. If number is divided, it is not prime number.
Else, if number is one, it is not prime number. 
Then the rest is prime number.
To sum the prime numbers, sum starts from zero. This will prevent the incremental of the results when I run the code several times. If the number is prime number below 2000, sum the number.

Challenge 3: Multiples of 3 and 5
If we list all of the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6, and 9. The sum of these multiples is 23. Find the sum of all the multiples of 3 and 5 below 1,000. Afterward, write a brief explanation walking through your code's logic in markdown.

In [165]:
sum_multiple = 0
# Find the numbers of multiples of 3 or 5 below 1000
for m in range(1,1000):
    if m % 3 == 0 or m % 5 == 0:
        sum_multiple += m
# Sum multiples   
print (sum_multiple)

233168


Sum starts from zero.
Find the multple of 3 or 5 below 1000.
Then sum multiples. And print the results.

Challenge 4: String Compressor
Implement a method to perform basic string compression using the counts of repeated characters. (This is called run-length encoding.) For example, the string "aabcccccaaa" would become a2b1c5a3. If the “compressed” string would not become smaller than the original string, your method should return the original string. You can assume the string has only uppercase and lowercase letters (a–z). Specify whether your solution is case sensitive or case insensitive and what you would need to change to make it the other. Afterward, write a brief explanation walking through your code's logic in markdown.

In [166]:
input_1 = "hhhhappppy"
input_2 = "abcd"
def res (input_str):
    comp_str = ""
    count_num = 1
     #input_str= input_str.lower()
    for i in range(len(input_str)-1):
        if input_str[i] == input_str[i+1]:
            count_num += 1
        else: 
            comp_str += input_str[i] + str(count_num)
            count_num = 1
    comp_str += input_str[i] + str(count_num)

    if len(comp_str) >= len(input_str):
        return input_str
    else:
        return comp_str
    
print(res(input_1))
print(res(input_2))    

h4a1p4p1
abcd


Suppose there are lower case charcaters. It is case-sensitive.
Starting from first character using loop, I will count it as 1 (count_num=1). Then will look at the next character. If it is equal to the previous character, I will add 1. If they are not equal, I will write the charater and 1. Append the comprssed string to comp_str efore moving to the next charactor.
Then, compare the length of the original string and compresed string. Show the string with the smaller number of characters,
Code is case sensitive. so by addin input_str= input_str.lower(), upper case string can be changed to lower case befor loop stars.

BONUS Challenge: FizzBuzz
Write a program that prints all of the numbers from 1 to 100. For multiples of 3, instead of the number, print "Fizz;" for multiples of 5, print "Buzz." For numbers that are multiples of both 3 and 5, print "FizzBuzz." Afterward, write a brief explanation walking through your code's logic in markdown.


In [167]:
numfb =range(1,100)
for fb in numfb:
    if fb % 5 == 0 and fb % 3 == 0:
        print("FizzBuzz")
    elif fb % 3 == 0:
        print("Fizz")
    elif fb % 5 == 0:
         print("Buzz")
    else:
        print(fb)

1
2
Fizz
4
Buzz
Fizz
7
8
Fizz
Buzz
11
Fizz
13
14
FizzBuzz
16
17
Fizz
19
Buzz
Fizz
22
23
Fizz
Buzz
26
Fizz
28
29
FizzBuzz
31
32
Fizz
34
Buzz
Fizz
37
38
Fizz
Buzz
41
Fizz
43
44
FizzBuzz
46
47
Fizz
49
Buzz
Fizz
52
53
Fizz
Buzz
56
Fizz
58
59
FizzBuzz
61
62
Fizz
64
Buzz
Fizz
67
68
Fizz
Buzz
71
Fizz
73
74
FizzBuzz
76
77
Fizz
79
Buzz
Fizz
82
83
Fizz
Buzz
86
Fizz
88
89
FizzBuzz
91
92
Fizz
94
Buzz
Fizz
97
98
Fizz


Numbfb's range is specified from 1 to 100.
Loop will first mark FIZZBUZZ for numbers divided by both 3 and 5 as order of sequence matters. Then if number is not mutiples of 3 and 5, then code will check the multiples of 3 to print Fizz. Then code will check the multiples of 5 to print Buzz. If a number doesn't meet the condition above, it will just print out the number.
